### ISSUE TITLE
Integration tests for boolean and assertion field operations

### ISSUE THREAD

**Issue by grjte** [maticnetwork/miden#166]
Tests for boolean and assertion field operations need to be added to the [integration tests](https://github.com/maticnetwork/miden/blob/next/miden/tests/integration/operations/field_ops.rs).

Tests should follow the existing pattern with the following tests added for each of the boolean field operations and the `assert` operation:
- a test of simple success cases
- a test with applicable failure cases, if there are any (compilation and/or execution failure)
- ~a proptest of the expected success case~ (removed as it's not necessary)

The [u32 bitwise tests](https://github.com/maticnetwork/miden/blob/next/miden/tests/integration/operations/u32_ops/bitwise_ops.rs) provide an example.

Specs for the operations are in the [Miden Assembly note](https://hackmd.io/YDbjUVHTRn64F4LPelC-NA).

The field operations which are missing tests are:

- [ ] assert
- [ ] not
- [ ] and
- [ ] or
- [ ] xor

More information can also be seen in the [field_ops assembly parsers](https://github.com/maticnetwork/miden/blob/next/assembly/src/parsers/field_ops.rs).

**Comment by aga7hokakological** [comment]
I would like to take this issue if no one else is working on this.

**Comment by grjte** [comment]
@aga7hokakological that would be great! Please go for it

**Comment by aga7hokakological** [comment]
My question is regarding the following points: 

> a test with applicable failure cases, if there are any (compilation and/or execution failure)

Here in below image given failure condition for all the operations which requires test. So does that failure case comes under the above point?

![image](https://user-images.githubusercontent.com/47058975/166456798-de1b228a-e964-46ff-a52a-16a5bc578b4e.png)

**Comment by grjte** [comment]
@aga7hokakological I just wanted to check in and see if you have any other questions about this task that I can help with. It's also fine if it turns out you don't have time to work on it - please just let us know if that's the case :)

**Comment by aga7hokakological** [comment]
Hey @grjte sorry, I was away for a while. I had figured out about the success cases but for failed cases such as for `not` condition failes if a > 1. How am I supposed to write the case. because it takes only binary input so no values other than 0 and 1 are allowed.

**Comment by grjte** [comment]
Hey @aga7hokakological no problem. Is the problem that it's throwing a compilation error so it never gets a chance to execute and therefore an execution error is never triggered? If that's the case, I think it's fine to just test for a TestError::AssemblyError here. If that's not the issue can you give me a bit more detail?

**Comment by aga7hokakological** [comment]
@grjte So here,
![image](https://user-images.githubusercontent.com/47058975/168803851-86b501b9-2852-459e-af5d-54e5e6c39035.png)

If I do `ExecutionError` it gives `Some(false) != Some(true)` and value is not binary value.
![image](https://user-images.githubusercontent.com/47058975/168804192-0123c121-363c-4e21-8979-00bcf0a80c89.png)

Here ExecutionError is required but as it is not binary value test is failing. 

For `AssemblyError` it gives `None` and `Some(right)` 
![image](https://user-images.githubusercontent.com/47058975/168804539-a3bcf761-7900-437e-8892-337250c40414.png)

**Comment by grjte** [comment]
@aga7hokakological Ok, I see. I think the issue is that you're using the wrong input to your`TestError::ExecutionError`. The input should be a substring that will match the expected error returned by the VM. In this case, the expected error is "NotBinaryValue", so if you do the following it should work:

```
build_op_test!(asm_op, &[2]).expect_error(TestError::ExecutionError("NotBinaryValue"));
```

**Comment by 0xKanekiKen** [comment]
> Test for `assert` works.
> For `assert.eq` what is opcode? `assert_eq` does not work.

The opcode for assert.eq is `assert.eq`. You can either refer it from [here](https://github.com/maticnetwork/miden/blob/next/assembly/src/parsers/field_ops.rs) (mentioned in the issue statement) or [here](https://hackmd.io/@bobbinth/ry-OIBwPF)

**Comment by grjte** [comment]
Thanks @0xKanekiKen! Yep, that's correct - in fact, the contents of the assembly note have been migrated to our docs now, so you can check out the assembly docs here: https://maticnetwork.github.io/miden/user_docs/assembly/main.html

@aga7hokakological regarding the first question about the proptests:
- I think using bool and casting to u64 as you've done would be the right way to do this.
- You don't need to consider the Felt data type, because that's handled inside the test handler. The test just takes in integers for both inputs and expected result.

You've brought up a good point though - since the only valid inputs are boolean values, proptests are probably overkill here, and we should probably stick to unit tests, since there are only 2 or 4 options to test for each of these anyway. We use them for all other tests which is why I had them specified in this issue, but I shouldn't have - sorry for my mistake and the extra work

**Comment by aga7hokakological** [comment]
#210

**Comment by aga7hokakological** [comment]
> you can check out the assembly docs here: https://maticnetwork.github.io/miden/user_docs/assembly/main.html

Thank you for this. Below code helped me.

> You can either refer it from [here](https://github.com/maticnetwork/miden/blob/next/assembly/src/parsers/field_ops.rs) (mentioned in the issue statement) or [here](https://hackmd.io/@bobbinth/ry-OIBwPF)

PR: #210 
I made mistake while committing to my branch so I didn't mentioned the PR directly in here.

**Comment by grjte** [comment]
Closed by #210
